internal/querycoord/global_meta_broker.go

```diff
@@ -267,41 +269,152 @@ func (broker *globalMetaBroker) parseIndexInfo(ctx context.Context, segmentID Un
+	indexCodec := storage.NewIndexFileBinlogCodec()
+	for _, indexFilePath := range fieldPathInfo.IndexFilePaths {
+		// get index params when detecting indexParamPrefix
+		if path.Base(indexFilePath) == storage.IndexParamsKey {
+			content, err := broker.cm.MultiRead([]string{indexFilePath})
+			if err != nil {
+				return nil, err
+			}
+
+			if len(content) <= 0 {
+				return nil, fmt.Errorf("failed to read index file binlog, path: %s", indexFilePath)
+			}
+
+			indexPiece := content[0]
+			_, indexParams, indexName, _, err := indexCodec.Deserialize([]*storage.Blob{{Key: storage.IndexParamsKey, Value: indexPiece}})
+			if err != nil {
+				return nil, err
+			}
+
+			return &extraIndexInfo{
+				indexName:   indexName,
+				indexParams: funcutil.Map2KeyValuePair(indexParams),
+			}, nil
+		}
+	}
+	return nil, errors.New("failed to load index extra info")
+}
+
+func (broker *globalMetaBroker) describeSegments(ctx context.Context, collectionID UniqueID, segmentIDs []UniqueID) (*rootcoordpb.DescribeSegmentsResponse, error) {
+	resp, err := broker.rootCoord.DescribeSegments(ctx, &rootcoordpb.DescribeSegmentsRequest{
+		Base: &commonpb.MsgBase{
+			MsgType: commonpb.MsgType_DescribeSegments,
+		},
+		CollectionID: collectionID,
+		SegmentIDs:   segmentIDs,
+	})
+	if err != nil {
+		log.Error("failed to describe segments",
+			zap.Int64("collection", collectionID),
+			zap.Int64s("segments", segmentIDs),
 			zap.Error(err))
 		return nil, err
 	}
-	indexInfo := &querypb.VecFieldIndexInfo{
-		FieldID: vecFieldIDs[0],
-	}
-	// check the buildID of the segment's index whether exist on rootCoord
-	enableIndex, buildID, err := broker.getIndexBuildID(ctx, collectionID, segmentID)
+
+	log.Debug("describe segments successfully",
+		zap.Int64("collection", collectionID),
+		zap.Int64s("segments", segmentIDs))
+
+	return resp, nil
+}
+
+// return: segment_id -> segment_index_infos
+func (broker *globalMetaBroker) getFullIndexInfos(ctx context.Context, collectionID UniqueID, segmentIDs []UniqueID) (map[UniqueID][]*querypb.VecFieldIndexInfo, error) {
+	resp, err := broker.describeSegments(ctx, collectionID, segmentIDs)
 	if err != nil {
 		return nil, err
 	}
 
-	// if the segment.EnableIndex == false, then load the segment immediately
-	if !enableIndex {
-		indexInfo.EnableIndex = false
-	} else {
-		indexInfo.BuildID = buildID
-		indexInfo.EnableIndex = true
-		err = broker.parseIndexInfo(ctx, segmentID, indexInfo)
-		if err != nil {
-			return nil, err
+	ret := make(map[UniqueID][]*querypb.VecFieldIndexInfo)
+	for _, segmentID := range segmentIDs {
+		infos, ok := resp.GetSegmentInfos()[segmentID]
+		if !ok {
+			log.Warn("segment not found",
+				zap.Int64("collection", collectionID),
+				zap.Int64("segment", segmentID))
+			return nil, fmt.Errorf("segment not found, collection: %d, segment: %d", collectionID, segmentID)
+		}
+
+		if _, ok := ret[segmentID]; !ok {
+			ret[segmentID] = make([]*querypb.VecFieldIndexInfo, 0, len(infos.IndexInfos))
```

Comment:
When building an index, each field should be limited to only one index, otherwise the load may fail

Reply:
We use `EnableIndex` to control this.

Reply:
It seems that we missing this control in rootcoord. When multiple indexes are created on same field, the last index should be enable and the rest should be disable. I'll fix this later.